lib/Transforms/ConvertSecretForToStaticFor/ConvertSecretForToStaticFor.cpp

```diff
@@ -0,0 +1,223 @@
+          forOp.getLoc(), forOp->getAttrOfType<IntegerAttr>("lower").getInt());
+
+      newForOp.setLowerBound(newLowerConstantOp);
+
+      // Set secretness for newForOp's lowerBound. Needed for bound checks in
+      // future RewritePattern calls on scf::ForOp
+      auto newLowerBound = newForOp.getLowerBound();
+      auto *newLowerBoundLattice =
+          solver->getOrCreateState<SecretnessLattice>(newLowerBound);
+      Secretness *newSecretness = &(newLowerBoundLattice->getValue());
+      newSecretness->setSecretness(false);
+
+      // Create CmpI Op
+      rewriter.setInsertionPointToStart(&(newForOp.getRegion().front()));
+      cmpILower = rewriter.create<arith::CmpIOp>(
+          newForOp.getRegion().getLoc(), arith::CmpIPredicate::sge,
+          inductionVariable, forOp.getLowerBound());
+      if (!isUpperBoundSecret) {
+        rewriter.setInsertionPointAfter(cmpILower);
+      }
+    }
+
+    // If upperBound is secret, rewrite (replace bound with respective attribute
+    // value and execute operations under scf.if that checks if i < upperBound)
+    if (isUpperBoundSecret) {
+      if (!forOp->getAttrOfType<IntegerAttr>("upper")) {
+        InFlightDiagnostic diag = mlir::emitWarning(
+            forOp.getLoc(),
+            "Cannot convert secret scf.for to static scf.for "
+            "since a static upper bound attribute has not been provided:");
+
+        return failure();
+      }
+
+      rewriter.setInsertionPoint(forOp);
+
+      auto newUpperConstantOp = rewriter.create<arith::ConstantIndexOp>(
+          forOp.getLoc(), forOp->getAttrOfType<IntegerAttr>("upper").getInt());
+
+      newForOp.setUpperBound(newUpperConstantOp);
+
+      // Set secretness for newForOp's upperBound. Needed for bound checks in
+      // future RewritePattern calls on scf::ForOp
+      auto newUpperBound = newForOp.getUpperBound();
+      auto *newUpperBoundLattice =
+          solver->getOrCreateState<SecretnessLattice>(newUpperBound);
+      Secretness *newSecretness = &(newUpperBoundLattice->getValue());
+      newSecretness->setSecretness(false);
+
+      // Create CmpI Op
+      if (isLowerBoundSecret)
+        rewriter.setInsertionPointAfter(cmpILower);
+      else
+        rewriter.setInsertionPointToStart(&(newForOp.getRegion().front()));
+
+      cmpIUpper = rewriter.create<arith::CmpIOp>(
+          newForOp.getRegion().getLoc(), arith::CmpIPredicate::slt,
+          inductionVariable, forOp.getUpperBound());
+
+      if (!isLowerBoundSecret) {
+        rewriter.setInsertionPointAfter(cmpIUpper);
+      }
+    }
+
+    // If both lowerBound and upperBound are secret, join the CmpIs with and
+    // AndI Op
+    if (isLowerBoundSecret && isUpperBoundSecret) {
+      // Create AndI Op
+      rewriter.setInsertionPointAfter(cmpIUpper);
+      andI = rewriter.create<arith::AndIOp>(cmpIUpper.getLoc(), cmpILower,
+                                            cmpIUpper);
+      rewriter.setInsertionPointAfter(andI);
```

Comment:
This should be a no-op since the insertion point is always moved to after the most recently created op. Same a few places elsewhere.